### PR TITLE
Improve message dispatching & small refactorings

### PR DESCRIPTION
**Pull request by jjj-vtm**
Hi, 

I made some small improvements dispatching messages to a listener. Basically:
- Drop locks early, especially drop them before the dispatch ie. on_receive call
- Dispatch is done on a separate task so that a slow listener does not slow down. 
- Todo to handle the connection drop case correctly.

Cheers, 

Jan

### CODE REVIEW

**Comment by PLeVasseur** [comment]
Hey @jjj-vtm -- sorry, I missed this! I'll make some time to review

**Comment by jjj-vtm** [comment]
No problem, I merged the changes.

**Review by PLeVasseur — COMMENTED** [review]
Hey @jjj-vtm -- thanks for this contribution! I left a few comments. Can you take a look?

File: src/lib.rs [hunk 1]
```diff
@@ -398,20 +396,33 @@ impl UPClientMqtt {
...
-                    for (_topic, listeners) in topics_iter {
-                        for listener in listeners.iter() {
-                            listener.on_receive(umessage.clone()).await;
-                        }
+                        .filter(|(key, _)| UPClientMqtt::compare_topic(topic, key))
+                        .flat_map(|(_topic, listener)| listener.to_owned())
+                        .collect();
+                    // Drop lock before message dispatch
+                    drop(topic_map_read);
+
```
[the remark this answers is not in the document]
Comment (jjj-vtm) [reply src/lib.rs:420]: Missed that ... which proves the point below that explicit dropping can easily be forgotten.

File: src/lib.rs [hunk 2]
```diff
@@ -398,20 +396,33 @@ impl UPClientMqtt {
                         trace!("No listeners registered for topic: {}", sub_topic);
                         continue;
                     };
+                    let owned_listeners = listeners.clone();
+
+                    // Release the locks early, before message dispatch
+                    drop(topic_map_read);
```
[the remark this answers is not in the document]
Comment (jjj-vtm) [reply src/lib.rs]: Yes, that is the way to go. I put the lock acquisition into their own scopes and factored out the actual notification.

**Pull request merged**